### ISSUE TITLE
The initial implicit argument names can differ from the specified names

### ISSUE THREAD

**Issue by jashug** [coq/coq#6785]
#### Version

8.8 master hash 8dd6d09 and 8.7.0

#### Description of the problem

```
Definition foo (a a : nat) : nat := 3.
Fail Arguments foo a a.
(*
To rename arguments the "rename" flag must be specified.
Argument a0 renamed to a.
*)

Fail Arguments foo a a : rename.
(* Some argument names are duplicated: a *)

Arguments foo a a0 : assert.
Arguments foo a b : rename.
```

This behavior seems confusing and fragile. Would it be possible to allow repeated names for argument specifications, and use the first one on an implicit application `(a := 2)`? As opposed to actual binders, I don't see a problem with shadowing uses in later types. One issue I do see is that queries like `About` and `Print` will show the renamed varaibles, but that is the same situation as after an explicit `Arguments foo c d`, for which `About` gives
```
foo : nat -> forall a0 : nat, a0 = a0
Arguments are renamed to c, d
...
```
Having instead ```Arguments are named a, a``` always printed doesn't seem too different.

---

One effect of the option in #6582 is to also mangle default argument names, since they get refreshed (with respect to the local env. not the global env.). This caused problems working on #6781.
I think that changing the above behavior would be an improvement to the system, making it easier to understand and work with, but the argument could also be made that #6582 is too agressive here.

**Comment by jashug** [comment]
It's not even impossible to get into this situation today:
```
Definition foo {arg_2 _ : nat} : nat := 42.
About foo.
(*
foo : nat -> nat -> nat
Arguments arg_2, arg_2 are implicit and maximally inserted
*)
Definition bar := foo (arg_2 := 2) (arg_2 := 3).
(* Argument name arg_2 occurs more than once. *)
```

The implicit argument mechanism chooses `arg_n` for the name of the n-th implicit argument if it is anonymous.

**Comment by jashug** [comment]
But `Arguments baz a {_} : rename.` fails with `Argument _ cannot be declared implicit.`?

**Comment by jashug** [comment]
```
Definition foo {x : nat} (y : nat) {z : nat} (w : nat) := 3.
Definition bar := foo (z := 7) 3 2 (x := 2).
```

After reading the code and realizing that the above works, I'd be happy with treating default implicit argument names that would be alpha-renamed as having no name, i.e. impossible to use the (x := x') argument syntax with. Thus if you don't repeat names, you are guaranteed to get the names you asked for, and in any case you never get names you didn't ask for.

**Comment by jashug** [comment]
https://github.com/coq/coq/blob/4f65dfb13d8bb395abf4aa405cae9ed529302a06/vernac/vernacentries.ml#L1141-L1148

**Comment by jashug** [comment]
So, I think that implicit arguments should not need to be named, and that the named argument feature can and should be separated from it. Then implicit arguments should just be a list of either explicit, or implicit with maximal implicit and force inference flags. #6838 does some initial refactoring to make changes easier.

**Comment by maximedenes** [comment]
I fully support that view. Will be glad to review the PRs as soon as I find a bit of time.

**Comment by herbelin** [comment]
> So, I think that implicit arguments should not need to be named, and that the named argument feature can and should be separated from it. Then implicit arguments should just be a list of either explicit, or implicit with maximal implicit and force inference flags. #6838 does some initial refactoring to make changes easier.

@jashug, @maximedenes: what does your proposal give in practice, on the pathological examples given above, or on the examples given in #6786, etc.?

**Comment by jashug** [comment]
I see two options, and don't have much of an opinion between them:
1. Allow repeated names for named arguments, apply explicit applications from left to right (as suggested in my first comment).
2. Don't perform (delta/any) reduction, warn on repeated names, and choose some way of disambiguating (so `a a a`  would become `a _ _` or `_ _ _` or `_ _ a`).
In either case, the `Arguments` command can be used to change names after the definition.

-----

`Definition foo {arg_2 _ : nat} : nat := 42.` would get names `arg_2 _`.
Here both arguments are maximal implicit.

`Definition foo (a a : nat) : nat :=3.`:
Under option 1, gets names `a a`, under option 2, gets a warning and names `a _`.
Neither argument is implicit.

Why we shouldn't delta reduce in option 2:
```
Definition foo (x : nat) := forall (y : nat), x = y.
Definition bar : forall (y : nat), foo y := _.
```
Delta reducing the type of `bar` becomes `forall (y y : nat), ^y = y` (where I used a carret for deBruijn indexing), which has repeated names `y y`. Instead, `bar` should get names `y _` in option 2, with no warning.
Neither argument is implicit, though `Set Implicit Arguments. Set Contextual Implicit.` would make both maximal implicit.

`Fixpoint foo {arg_1 arg_1 _ : nat} := _.`:
Under option 1 would get names `arg_1 arg_1 _`, under option 2 would get a warning and names `arg_1 arg_1 _`.
All three arguments are maximal implicit.

After a statement is defined, you can always use `Arguments` to rename names, so we don't need to work too hard to make the default names comfortable. But before it is defined, in a `Fixpoint` or `Inductive`, especially with `Set Implicit Arguments`, there is no good way to find the default names, so you want them to follow a simple set of rules. (Similarly, with `Set Implicit Arguments`, it applies to uses inside a `Fixpoint` or `Inductive`, and you can't figure out which arguments have been made implicit. I almost feel I'd like the effect of `Set Implicit Arguments` to be equivalent to doing `Arguments foo : default implicits.` after declaration of `foo`, therefore not applying to to body of `foo` itself, but that's another issue).

**Comment by jashug** [comment]
Annother way to get repeated argument names:
```
Section foo.
Context {a : nat}.
Definition bar (e : a = a) {a : nat} := a.
End foo.
About bar.
(*
bar : forall a : nat, a = a -> nat -> nat
Arguments a, a are implicit and maximally inserted
*)
```

**Comment by jashug** [comment]
A related issue: #4696

**Comment by herbelin** [comment]
@jashug:

I would try to be even stricter and to fail in case of duplicated names in the "toplevel spine" of binders in type of a definition or the statement of a theorem, as `Arguments` already does when given two identical names. Idem in the presence of section variables: if a variable name is already used as a section variable name, it cannot be reused in the toplevel spine of a definition/theorem. Also potentially taking into account some toplevel `Eval some-reduction in` the definition.

As for the automatically named `arg_n` maximal arguments, I don't remember what was the motivation for naming them. If the motivation was to be able to use the `f (arg_n:=t)` notation, one could alternatively provide a `f (2:=t)` notation (it is not a very elegant notation, e.g. `f (1st:=t)`, `f (2nd:=t)`, `f (3rd:=t)`, etc. could indicate more explicitly that it is about positions that we are talking :wink:, but it is at least consistent with the `(2:=t)` notation used in `apply with`).

So, shortly: if nothing requires to have an explicit name, I guess it is ok to keep them "unnamed" as originally expected by the user.

Another question in passing: is there any reason to restrict the `(x:=t)` notation to only implicit arguments, i.e. not to potentially use names as labels?
 
Failing would also prevent to swing between the two non-ideal ways to solve #307-style situations: 
```coq
Lemma f (x:nat) : forall y, x=y.
refine (fun x => _).
(* x0 : nat  x : nat |- x0 = x *)
```
(Failing was suggested in the #307 discussion.)

**Comment by jashug** [comment]
> Another question in passing: is there any reason to restrict the (x:=t) notation to only implicit arguments, i.e. not to potentially use names as labels?

No, it seems sensible to support `(x := y)` even if `x` is not implicit.

> I would try to be even stricter and to fail in case of duplicated names in the "toplevel spine" of binders in type of a definition or the statement of a theorem

What do you think the initial names of the arguments of `foo` should be in the following?
```
Axiom X : Set.
Definition commutative (f : X -> X -> X) := forall x y, f x y = f y x.
Axiom foo : forall (x : X -> X -> X), commutative x.
```
If you delta-expand `commutative` (as the implicit argument system currently does), you get repeated names `x`.

Not allowing shadowing definitely makes terms easier to reason about (with named representations). But then the set of visible names becomes part of the context; I worry about having to do more manual alpha-renaming.

**Comment by jashug** [comment]
Issue #9098 may be fixed at the same time, recording here so I don't forget about it.
